pyra/hardware.py

```diff
@@ -0,0 +1,424 @@
+"""
+..
+   hardware.py
+
+Functions related to the dashboard viewer.
+"""
+# standard imports
+import json
+
+# lib imports
+import GPUtil
+import plotly
+import psutil
+
+# local imports
+from pyra import definitions
+from pyra import helpers
+from pyra import locales
+from pyra import logger
+
+_ = locales.get_text()
+chart_translations = dict(
+    cpu=dict(
+        bare=_('cpu'),
+        usage=_('cpu usage')
+    ),
+    gpu=dict(
+        bare=_('gpu'),
+        usage=_('gpu usage')
+    ),
+    memory=dict(
+        bare=_('memory'),
+        usage=_('memory usage')
+    ),
+    network=dict(
+        bare=_('network'),
+        usage=_('network usage')
+    ),
+    general=dict(
+        received=_('received'),
+        sent=_('sent'),
+        system=_('system')
+    )
+)
+
+log = logger.get_logger(__name__)
+
+initialized = False
+network_recv_last = 0
+network_sent_last = 0
+proc = psutil.Process()  # the main retroarcher process
+proc_id = proc.pid
+processes = [proc]
+
+nvidia_gpus = GPUtil.getGPUs()
+
+try:
+    import pyamdgpuinfo
+except ModuleNotFoundError:
+    amd_gpus = 0
+else:
+    amd_gpus = pyamdgpuinfo.detect_gpus()  # this will be an integer representing the count of amd gpus
+
+dash_stats = dict(
+    timestamp=[],
+    relative_time=[],
+    cpu_system=[],
+    memory_system=[],
+    network_sent=[],
+    network_received=[]
+)
+
+history_length = 120
+
+
+def update_cpu() -> float:
+    """
+    Update dashboard stats for system CPU usage.
+
+    This will append a new value to the ``dash_stats['cpu_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system cpu percentage utilized.
+
+    Examples
+    --------
+    >>> update_cpu()
+    """
+    cpu_percent = min(100, psutil.cpu_percent(interval=None, percpu=False))  # max of 100
```

Comment:
Why clamp the max to 100? I'm accustomed to CPU percentages being over 100 in some cases.

This might be a good place to raise an exception.